framework/db/ActiveRelationTrait.php

```diff
@@ -380,21 +388,41 @@ private function populateInverseRelation(&$primaryModels, $models, $primaryName,
      * @param array $models
      * @param array $link
      * @param array $viaModels
-     * @param array $viaLink
+     * @param null|self $viaQuery
      * @param bool $checkMultiple
      * @return array
      */
-    private function buildBuckets($models, $link, $viaModels = null, $viaLink = null, $checkMultiple = true)
+    private function buildBuckets($models, $link, $viaModels = null, $viaQuery = null, $checkMultiple = true)
     {
         if ($viaModels !== null) {
             $map = [];
+            $viaLink = $viaQuery->link;
             $viaLinkKeys = array_keys($viaLink);
             $linkValues = array_values($link);
             foreach ($viaModels as $viaModel) {
                 $key1 = $this->getModelKey($viaModel, $viaLinkKeys);
                 $key2 = $this->getModelKey($viaModel, $linkValues);
                 $map[$key2][$key1] = true;
             }
+
+            $viaQuery->viaMap = $map;
+
+            $viaVia = $viaQuery->via;
+            while ($viaVia) {
+                $viaViaQuery = is_array($viaVia) ? $viaVia[1] : $viaVia;
+                $viaMap = $viaViaQuery->viaMap;
+                $oldMap = $map;
+                $map = [];
+                foreach ($oldMap as $key1 => $keys) {
```

Comment:
Don't like this way, key1, key2, key3, a lot of foreaches, can we do it better ?

Reply:
I'll try to change